base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs

```diff
@@ -3132,59 +3133,78 @@ fn run_migrations(db: &mut LMDBDatabase) -> Result<(), ChainStorageError> {
-                            height
+                            "[MIGRATIONS] v{}: Block height {} already has correct accumulated difficulty",
+                            migrate_from_version, height
                         );
                         last_correct_height = height;
                     }
                 } else {
-                    info!(target: LOG_TARGET, "No accumulated difficulty found for block height {}", height);
+                    info!(
+                        target: LOG_TARGET,
+                        "[MIGRATIONS] v{}: No accumulated difficulty found for block height {}",
+                        migrate_from_version, height
+                    );
                     break;
                 }
             }
             if last_correct_height == 0 {
                 // this will happen only happen if the db is below the fork height of the RxT fork
-                info!(target: LOG_TARGET, "No migration to perform for version network");
+                info!(
+                    target: LOG_TARGET,
+                    "[MIGRATIONS] v{}: No migration to perform for version network",
+                    migrate_from_version
+                );
                 continue;
             }
             // lets rewind to last known good accumulated difficulty so the db can be correctly calculated again
             rewind_to_height(db, last_correct_height)?;
         }
-        if migrate_from_version == 2 {
-            // Verify database consistency before starting migration
-            info!(target: LOG_TARGET, "Starting PayRef migration");
+
+        // MIGRATION: Add payref index, rebuild payref index to recover deleted payrefs, add reverse payref index
+        // Note: For previously running base nodes `last_migrated_version` was incremented beyond `MIGRATION_VERSION` up
+        //       to `migrate_from_version == 4`. This migration also fix the error introduced with the original payref
+        //       migration where it was only added for outputs in the unspent set, resulting in missing payrefs.
+        if (migrate_from_version == 2 || migrate_from_version == 3 || migrate_from_version == 4) && !payref_index_done {
+            info!(target: LOG_TARGET, "[MIGRATIONS] v{}: Starting PayRef migration", migrate_from_version);
 
             let read_txn = db.read_transaction()?;
-            let chain_height = match fetch_chain_height(&read_txn, &db.metadata_db) {
-                Ok(v) => v,
-                Err(_) => {
-                    // No chain data, skip PayRef rebuild
-                    drop(read_txn);
-                    continue;
-                },
-            };
+            let chain_height = fetch_chain_height(&read_txn, &db.metadata_db).unwrap_or(0);
             drop(read_txn);
-            // we need to clear the payref index first as they might now bw wrong
+
+            // Clear the payref index first as they might now be wrong
```

Comment:
we dont need to delete it

Reply:
Agreed, a remnant of the previous code - replace will update it anyway.